smack-extensions/src/main/java/org/jivesoftware/smackx/usertune/UserTuneManager.java

```diff
@@ -0,0 +1,167 @@
+import org.jivesoftware.smack.SmackException.NotConnectedException;
+import org.jivesoftware.smack.SmackException.NotLoggedInException;
+import org.jivesoftware.smack.XMPPConnection;
+import org.jivesoftware.smack.XMPPException.XMPPErrorException;
+import org.jivesoftware.smack.packet.Message;
+import org.jivesoftware.smack.provider.ProviderManager;
+import org.jivesoftware.smackx.disco.ServiceDiscoveryManager;
+import org.jivesoftware.smackx.pep.PepListener;
+import org.jivesoftware.smackx.pep.PepManager;
+import org.jivesoftware.smackx.pubsub.EventElement;
+import org.jivesoftware.smackx.pubsub.ItemsExtension;
+import org.jivesoftware.smackx.pubsub.LeafNode;
+import org.jivesoftware.smackx.pubsub.PayloadItem;
+import org.jivesoftware.smackx.pubsub.PubSubException.NotALeafNodeException;
+import org.jivesoftware.smackx.pubsub.PubSubManager;
+import org.jivesoftware.smackx.usertune.element.UserTuneConcretisation;
+import org.jivesoftware.smackx.usertune.element.UserTuneElement;
+import org.jivesoftware.smackx.usertune.provider.UserTuneConcretisationProvider;
+
+import org.jxmpp.jid.BareJid;
+import org.jxmpp.jid.EntityBareJid;
+
+/**
+ * Entry point for Smacks API for XEP-0118: User Tune.
+ * <br>
+ * To set a usertune, please use one of the {@link #setUserTune(UserTune)} methods. This will publish the usertune to a pubsub
+ * node.<br>
+ * <br>
+ * In order to get updated about other user's tunes, register a {@link UserTuneListener} at
+ * {@link #addUserTuneListener(UserTuneListener)}. That listener will get updated about any incoming tune updates of contacts.<br>
+ * <br>
+ * To stop publishing the userhttps://xmpp.org/extensions/xep-0118.html#protocol-elementstune, refer to {@link #clearUserTune()}.<br>
+ * <br>
+ * It is also possible to add {@link UserTuneElement}s to {@link Message}s by using {@link #addUserTuneToMessage(Message, UserTune)}.<br>
+ * <br>
+ * The API can be extended with custom userTune concretisations by extending {@link UserTuneConcretisation} and registering
+ * {@link UserTuneConcretisationProvider}s using {@link ProviderManager#addExtensionProvider(String, String, Object)}.<br>
+ * <br>
+ * @see <a href="https://xmpp.org/extensions/xep-0118.html">
+ *     XEP-0118: User Tune</a>
+ */
+public final class UserTuneManager extends Manager {
+
+    public static final String USERTUNE_NODE = "http://jabber.org/protocol/tune";
+    public static final String USERTUNE_NOTIFY = USERTUNE_NODE + "+notify";
+
+    private static final Map<XMPPConnection, UserTuneManager> INSTANCES = new WeakHashMap<>();
+
+    private final Set<UserTuneListener> userTuneListeners = new HashSet<>();
+    private final AsyncButOrdered<BareJid> asyncButOrdered = new AsyncButOrdered<>();
+    private PubSubManager pubSubManager;
+
+    private UserTuneManager(XMPPConnection connection) {
+        super(connection);
+        ServiceDiscoveryManager.getInstanceFor(connection).addFeature(USERTUNE_NOTIFY);
```

Comment:
I'd swap this line and the line below. Even if it is super unlikely, that way you wont lose any incoming `UserTune` events that are received before you start listening for them.

Reply:
Arranged as per the suggestion.